dev-docs/RFCs/v8.0/luma-refactor.md

```diff
@@ -0,0 +1,83 @@
+- A large code base. ~733kB of source files - compare to e.g. a full 3D engine like three.js at ~1.2MB (1.6x), a low-level library like PicoGL at ~183kB (0.25x)
+- Several unfinished or unused APIs that make it hard to tell what's reliable
+- Documentation is sprawling and not focused on the API. Examples demonstrate random scenes rather than API features
+
+Clearly defining luma.gl's goals as a framework, and structuring the codebase and documentation around those goals, would benefit it in the following ways:
+- Reduce its complexity and associated maintenance costs
+- Reduce the weight it imposes on applications that use it
+- Make its value proposition clear to potential users
+- Simplify adoption
+
+## Overview
+
+The first step towards simplifying luma.gl is to be explicit about what its goals are. I propose that changes be made around the following key principles:
+- Targeting intermediate to advanced WebGL users
+- All systems implemented should solve explicitly identified problems. Experiments are encouraged, but those that fail to find a use case should be removed
+- Structured around levels of abstraction, primarily helper tools for low-level WebGL work, WebGL wrapper classes, and higher-level engine features
+- Increased focus on architectural fundamentals, simplicity, building systems that are easy to reason about. Less focus on using the latest APIs or frameworks
+
+
+The actual refactor of luma.gl would occur in three phases:
+
+### Clean Out Unused Code
+
+This will warrant some discussion. The following are systems I believe should be considered for removal or consolidation:
+- `SceneGraphNode`, `BaseModel`, `Model` - functionality could be supported with just the `Model` and `Group` classes
+- `CameraNode` - Appears unused
+- Uniform animations - Complex and unused
+- `ShaderCache` - Superseded by `ProgramManager`
+- core/lighting, core/materials - currently empty or simple data classes
+- core/multipass - Appears unused
+- seer integration - Complex and will interfere with other parts of the system if enabled (e.g. GPU timing). Is anyone using it?
+- debug/parameter-definitions - Appears unused
+- main module - Appears unused
+- core/geometry - Perhaps better suited to math.gl?
+
+### Structure Modules Around Meaningful Themes
+
+I propose reducing the number of modules to five, around the following themes:
+
+#### Constants
+
+The `constants` module would remain as it is now.
+
+#### Low-level
+
+luma.gl would would contain two low-level modules:
+- `shader-tools`: Current ShaderTools and shader debugging tools from current debug module
+- `context-tools`: Current webgl-polyfill and webgl-state-tracker, and context debugging tools from current debug module
+
+#### Mid-level
+
+The `webgl` module would remain as it is now
+
+#### High-level
+
+A new `engine` module would contain the higher-level constructs in luma.gl including models, resource management, rendering, GPGPU, etc.
```

Comment:
What's the reason to abandon the `core` module name? This module directly interfaces with deck.gl and I would like to keep breaking changes minimal.

Reply:
Assume we will have `core`+`engine`?

Reply:
Just that I don't consider core to be "core" in this structure. If anything the low-level modules are "core". We can discuss this though if it's considered problematic or too much of a deviation from other repos.